### ISSUE TITLE
wait is not working for already exited / finished / terminated jobs

### ISSUE THREAD

**Issue by mamiu** [fish-shell/fish-shell#7415]
fish version: `3.1.0 / 3.0.2`

Builtin `wait` command is not working for already exited / finished / terminated jobs.
See here: https://github.com/fish-shell/fish-shell/issues/1422#issuecomment-169410032

### Working:

```fish
$ sleep 10 &
$ jobs -lp | read JOB_PID
$ wait $JOB_PID
Job 1, 'sleep 10 &' has ended
```

### Not working:

```fish
$ sleep 10 &
$ jobs -lp | read JOB_PID
Job 1, 'sleep 10 &' has ended
$ wait $JOB_PID
wait: Could not find a job with process id '47299'
```

Are there any plans to add this feature (or fix this bug)?

**Comment by faho** [comment]
See also #7210. 

In this case it would probably enough to just not print an error - we could just assume it has ended.

Also as a sidenote: You can just use $last_pid, no need for the weird `read JOB_PID` dance.

**Comment by mamiu** [comment]
@faho Thanks for the tip!

The thing is: I want to get the exit code (`$status`) of the background job with the specific PID after its termination. If I hadn't got that error message I would have assumed that `wait` is working like in bash.

**Comment by mamiu** [comment]
I found this [in the other issue](https://github.com/fish-shell/fish-shell/issues/7210#issuecomment-660696953) you've mentioned:

> I think wait does not provide the `$status` of the exited job, am I wrong?

Does that mean, that even if wait supports terminated jobs it doesn't provide what I was looking for. Or is there another way to achieve this?

**Comment by faho** [comment]
That's correct, `wait` does not currently return the $status of the job it waited on, and there currently is no way of getting it.

It's also tricky, because if you can wait on any job that ever existed, and we need to return its $status, and most jobs are never `wait`ed on, then that's basically a memory leak. I don't know how other shells handle this.

**Comment by mamiu** [comment]
Very interesting! Thanks for the insight. Indeed interesting to know how other shells handle this.

Would be nice if fish passed the `$status` of the finished job to the function with the `--on-job-exit` argument (if there's any).

**Comment by mqudsi** [comment]
The other problem is that job ids are not monotonic and are reused. Who knows which of the hundreds of job 3 you are waiting on?

Having monotonic job ids (even if they're eventually recycled) more akin to PIDs would solve a lot of weird issues, but I'm not sure how you guys feel about that.

**Comment by hexagonrecursion** [comment]
> monotonic job ids (even if they're eventually recycled) more akin to PIDs

According to [this](https://unix.stackexchange.com/questions/16883/what-is-the-maximum-value-of-the-process-id) PIDs are not monotonic. They do get reused.

**Comment by hexagonrecursion** [comment]
> Indeed interesting to know how other shells handle this.

bash https://www.gnu.org/software/bash/manual/html_node/Bash-Variables.html#index-CHILD_005fMAX

> CHILD_MAX
> 
>     Set the number of exited child status values for the shell to remember. Bash will not allow this value to be decreased below a POSIX-mandated minimum, and there is a maximum value (currently 8192) that this may not exceed. The minimum value is system-dependent.
>
